webcrypto/subtle_crypto.go

```diff
@@ -160,8 +160,47 @@ func (sc *SubtleCrypto) Digest(algorithm goja.Value, data interface{}) *goja.Pro
 //
 // The `keyUsages` parameter is an array of strings indicating what the key can be used for.
 func (sc *SubtleCrypto) GenerateKey(algorithm goja.Value, extractable bool, keyUsages []CryptoKeyUsage) *goja.Promise {
-	// TODO: implementation
-	return nil
+	promise, resolve, reject := sc.makeHandledPromise()
+
+	// 2.
+	// As the algorithm could either be a string or object, we export it to
+	// a interface{}, and let NormalizeAlgorithm() handle the rest.
+	normalizedAlgorithm, err := NormalizeAlgorithm(algorithm.Export(), OperationIdentifierGenerateKey)
+	if err != nil {
+		reject(err)
+		return promise
+	}
+
+	// 5.
+	go func() {
+		var value goja.Value
+		var err error
+
+		// FIXME: params could be an interface here, depending on whatever underlying
```

Comment:
To continue the discussion started in https://github.com/grafana/xk6-webcrypto/pull/23#discussion_r1034378372

To be honest, it's pretty clear right now that we need an additional interface, and it can have the name `Algorithm` since now this struct does almost nothing, so it can be an internal `algorithm` with only `name` property inside.

```golang
type Algorithm interface {
    Name() string
    GenerateKey(*goja.Runtime, []CryptoKeyUsage) (goja.Value, error)
}
```

And AES, HMAC, and so on should just implement this interface.

Reply:
Yes, I think you're right, I'll give it a try, and rebase the other PRs (encrypt, decrypt, and incoming import/export) to see if the ideas expands nicely and if it introduces breaking changes down the road 👍🏻

Reply:
A `KeyGenerator` interface now enforces `*KeyGenParams` structs to implement the method. The `GenerateKey` method casts the normalized algorithm to a `KeyGenerator`  and is much simplified.